compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEPadAndBufferize.cpp

```diff
@@ -0,0 +1,172 @@
+                    linalg::LinalgDialect>();
+  }
+  void runOnOperation() override;
+};
+
+void AMDAIEPadAndBufferizePass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  IREE::HAL::ExecutableVariantOp variantOp = getOperation();
+  ModuleOp innerModule = variantOp.getInnerModule();
+  for (func::FuncOp funcOp : innerModule.getOps<func::FuncOp>()) {
+    linalg::MatmulOp matmulOp;
+    funcOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+        [&](TilingInterface op) {
+          // Find the next matmul op if it does not have loops.
+          if (op.getLoopIteratorTypes().empty() || !isa<linalg::MatmulOp>(op))
+            return WalkResult::advance();
+          matmulOp = cast<linalg::MatmulOp>(op);
+          return WalkResult::interrupt();
+        });
+    if (!matmulOp) {
+      LLVM_DEBUG(llvm::dbgs() << "----- skip, no matmul op -----\n");
+      return;
+    }
+
+    IRRewriter rewriter(context);
+    linalg::LinalgPaddingOptions options;
+    SmallVector<Attribute> paddingValues;
+    auto i64Type = rewriter.getI64Type();
+    for (auto operand : matmulOp->getOperands()) {
+      auto type = dyn_cast<RankedTensorType>(operand.getType());
+      if (!type) {
+        funcOp->emitOpError("failed to lower workgroup count region");
+        return signalPassFailure();
+      }
+      Type elementType = type.getElementType();
+      if (isa<IntegerType>(elementType)) {
+        paddingValues.push_back(rewriter.getIntegerAttr(elementType, 0));
+      } else {
+        paddingValues.push_back(rewriter.getFloatAttr(elementType, 0));
+      }
+    }
+    options.paddingValues = paddingValues;
+    SmallVector<bool> packPaddings(matmulOp->getNumOperands(), true);
+    options.packPaddings = packPaddings;
+
+    SmallVector<int64_t> paddingDimensions;
+    for (unsigned i = 0, n = matmulOp.getNumLoops(); i < n; i++)
+      paddingDimensions.push_back(i);
+    options.paddingDimensions = paddingDimensions;
+
+    SmallVector<int64_t> padToMultipleOf(options.paddingDimensions.size(), 1);
+    options.padToMultipleOf = padToMultipleOf;
+    options.copyBackOp = linalg::LinalgPaddingOptions::CopyBackOp::LinalgCopy;
+
+    if (failed(applyPadAndConvertToDPS(rewriter, matmulOp, options))) {
+      funcOp->emitOpError("failed to apply pad");
+      return signalPassFailure();
+    }
+
+    // TODO: This part can be coded better instead of againg walking the FuncOp.
+    funcOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+        [&](TilingInterface op) {
+          // Find the next matmul op if it does not have loops.
+          if (op.getLoopIteratorTypes().empty() || !isa<linalg::MatmulOp>(op))
+            return WalkResult::advance();
+          matmulOp = cast<linalg::MatmulOp>(op);
+          return WalkResult::interrupt();
+        });
+    if (!matmulOp) {
+      LLVM_DEBUG(llvm::dbgs() << "----- skip, no matmul op -----\n");
+      return;
+    }
+    for (auto operand : matmulOp->getOperands()) {
```

Comment:
Wait, we shouldnt be bufferizing yet. We need to do more levels tile and distribute on tensors right?